### DISCUSSION TITLE
Add NotNullWhen attribute to IsInstanceValid argument

### DISCUSSION THREAD

**Discussion by Joy-less** [godotengine/godot-proposals#10418] [Scripting]
The code for `Node.IsInstanceValid` in C# is as follows, according to the reflector:
```cs
public static bool IsInstanceValid(GodotObject? instance)
{
    if (instance != null)
    {
        return instance.NativeInstance != IntPtr.Zero;
    }
    return false;
}
```

One common use case of `IsInstanceValid` is to check whether an instance is not null and not disposed.
However, if you use it this way:
```cs
Node? MyNode = new();
public override void _Ready() {
    if (!IsInstanceValid(MyNode)) {
        return;
    }
    GD.Print(MyNode.Name); // Dereference of a possibly null reference.
}
```

This can be resolved by adding a `NotNullWhenAttribute` to the `instance` argument.
This tells the C# compiler that if the method returns true, the instance cannot be null.
```cs
public static bool IsInstanceValid([NotNullWhen(true)] GodotObject? instance)
{
    if (instance != null)
    {
        return instance.NativeInstance != IntPtr.Zero;
    }
    return false;
}
```

**Comment by raulsntos** [comment]
Sounds good. Feel free to open a pull request to add the attribute :slightly_smiling_face:.

**Reply by Joy-less** [reply]
Opened: https://github.com/godotengine/godot/pull/95431
